feat: acceleration and transport layer

knzo25 opened the pull request:
## Description
This is part of the series of PR related to https://github.com/autowarefoundation/autoware.universe/issues/9722

List of PRs:
 - https://github.com/autowarefoundation/autoware.universe/pull/9454 (pointcloud preprocessing)
 - https://github.com/autowarefoundation/autoware.universe/pull/10298 (concatenation - template refactor)
 - https://github.com/autowarefoundation/autoware.universe/pull/10300 (concatenation - cuda version)
 - https://github.com/autowarefoundation/autoware.universe/pull/9453 (centerpoint)
 - transfusion (TODO)
 - https://github.com/autowarefoundation/autoware.universe/pull/9542 (OGM - the first implementation will be independent of the blackboard to ease the transition)
 - https://github.com/tier4/aip_launcher/pull/348 (aip_launcher)
 - https://github.com/autowarefoundation/sample_sensor_kit_launch/pull/111 (sample_sensor_kit_launch)

Depending on your machine and how many nodes are in a container, the following branch may also be required:
https://github.com/knzo25/launch_ros/tree/fix/load_composable_node
There seems to be a but in ROS where if you send too many services at once some will be lost and `ros_launch` can not handle that.

## Related links

**Parent Issue:**

- Link

<!-- ⬇️🟢
**Private Links:**

- [CompanyName internal link]()
⬆️🟢 -->

## How was this PR tested?
The sensing/perception pipeline was tested until centerpoint for TIER IV's taxi using the logging simulator.

## Notes for reviewers
The main branch that I used for development is [`feat/cuda_acceleration_and_transport_layer2`](https://github.com/knzo25/autoware.universe/tree/feat/cuda_acceleration_and_transport_layer2).
However, the changes were too big so I split the PRs. That being said, development, if any will still be on that branch (and then cherrypicked to the respective PRs), and the review changes will be cherrypicked into the development branch.

## Interface changes

An additional topic is added to perform type negotiation:
Example: `input/pointcloud` -> `input/pointcloud` and `input/pointcloud/cuda`

<!-- ⬇️🔴

### Topic changes

#### Additions and removals

| Change type   | Topic Type      | Topic Name    | Message Type        | Description       |
|:--------------|:----------------|:--------------|:--------------------|:------------------|
| Added/Removed | Pub/Sub/Srv/Cli | `/topic_name` | `std_msgs/String`   | Topic description |

#### Modifications

| Version | Topic Type      | Topic Name        | Message Type        | Description       |
|:--------|:----------------|:------------------|:--------------------|:------------------|
| Old     | Pub/Sub/Srv/Cli | `/old_topic_name` | `sensor_msgs/Image` | Topic description |
| New     | Pub/Sub/Srv/Cli | `/new_topic_name` | `sensor_msgs/Image` | Topic description |

### ROS Parameter Changes

#### Additions and removals

| Change type   | Parameter Name | Type     | Default Value | Description       |
|:--------------|:---------------|:---------|:--------------|:------------------|
| Added/Removed | `param_name`   | `double` | `1.0`         | Param description |

#### Modifications

| Version | Parameter Name   | Type     | Default Value | Description       |
|:--------|:-----------------|:---------|:--------------|:------------------|
| Old     | `old_param_name` | `double` | `1.0`         | Param description |
| New     | `new_param_name` | `double` | `1.0`         | Param description |

🔴⬆️ -->

## Effects on system behavior

Enabling this preprocessing in the launchers should provide a much reduced latency and cpu usage (at the cost of a higher GPU usage)

vividf committed:
feat: initialize parameters for concatenate

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
chore: add xx1 gen2 parameters

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
chore: fix aip x1 parameter

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
chore: fix x2 parameter

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
Merge branch 'tier4/universe' into feat/new_concatenate_node_parameter

knzo25 committed:
feat: implemented the launcher-side changes for the cuda preprocessing and transport layer

Signed-off-by: Kenzo Lobos-Tsunekawa <kenzo.lobos@tier4.jp>

knzo25 commented:
Note: this PR will be kept as a draft until https://github.com/tier4/aip_launcher/pull/337 it merged, at which point this PR will be reabsed

knzo25 committed:
Merge remote-tracking branch 'origin/tier4/universe' into feat/cuda_acceleration_and_transport_layer

Signed-off-by: Kenzo Lobos-Tsunekawa <kenzo.lobos@tier4.jp>

knzo25 committed:
feat: accepting variable number of crop boxes now

Signed-off-by: Kenzo Lobos-Tsunekawa <kenzo.lobos@tier4.jp>

knzo25 committed:
Merge branch 'tier4/universe' into feat/cuda_acceleration_and_transport_layer

Signed-off-by: Kenzo Lobos-Tsunekawa <kenzo.lobos@tier4.jp>

knzo25 commented:
The concatenated PR was merged, so after solving the conflicts and some recent changes, I am opening this PR :pray:

knzo25 committed:
Merge branch 'tier4/universe' into feat/cuda_acceleration_and_transport_layer

knzo25 committed:
chore: updated the launchers to reflect the newest changes in the nodes implementation

Signed-off-by: Kenzo Lobos-Tsunekawa <kenzo.lobos@tier4.jp>

knzo25 committed:
Merge branch 'tier4/universe' into feat/cuda_acceleration_and_transport_layer

Signed-off-by: Kenzo Lobos-Tsunekawa <kenzo.lobos@tier4.jp>

mojomex commented on aip_xx1_launch/launch/pointcloud_preprocessor.launch.py:
```diff
@@ -37,17 +37,32 @@ def launch_setup(context, *args, **kwargs):
-        parameters=[concatenate_and_time_sync_node_param],
-        extra_arguments=[{"use_intra_process_comms": LaunchConfiguration("use_intra_process")}],
-    )
+    if LaunchConfiguration("use_cuda_preprocessor").perform(context):
+        concat_component = ComposableNode(
+            package="autoware_cuda_pointcloud_preprocessor",
+            plugin="autoware::cuda_pointcloud_preprocessor::CudaPointCloudConcatenateDataSynchronizerComponent",
+            name="concatenate_data",
+            remappings=[
+                ("~/input/twist", "/sensing/vehicle_velocity_converter/twist_with_covariance"),
+                ("output", "concatenated/pointcloud"),
+                ("output/cuda", "concatenated/pointcloud/cuda"),
+            ],
+            parameters=[concatenate_and_time_sync_node_param],
+            # NOTE(knzo25): when using  the cuda blackboard, this setting can not be made global
+            # extra_arguments=[{"use_intra_process_comms": LaunchConfiguration("use_intra_process")}],
+        )
+    else:
+        concat_component = ComposableNode(
+            package="autoware_pointcloud_preprocessor",
+            plugin="autoware::pointcloud_preprocessor::PointCloudConcatenateDataSynchronizerComponent",
+            name="concatenate_data",
+            remappings=[
+                ("~/input/twist", "/sensing/vehicle_velocity_converter/twist_with_covariance"),
+                ("output", "concatenated/pointcloud"),
+            ],
+            parameters=[concatenate_and_time_sync_node_param],
+            extra_arguments=[{"use_intra_process_comms": LaunchConfiguration("use_intra_process")}],
+        )
```
The two branches differ only in their package, plugin and extra_params. The rest of the code could be merged into one to prevent future contributors from only modifying the one they're concerned about

knzo25 replied:
Addressed in [bfcf9a2](https://github.com/tier4/aip_launcher/pull/348/commits/bfcf9a2a563dadc3ef4d6c96e4dc60049314c392)

mojomex commented on common_sensor_launch/launch/nebula_node_container.launch.py:
This was already the case before this PR, but this launch file is spaghetti code, making it very hard to understand at a glance what is going on.

Especially with the kind of multiplexed code of CUDA/non-CUDA, I'd suggest factoring out into a few functions like these:

```python
def make_cuda_pointcloud_preprocessor(distortion_corrector_params, ring_outlier_params, vehicle_info, mirror_info) -> ComposableNode:
  ...

def make_vehicle_cropbox_filter(vehicle_info) ->ComposableNode:
  ...

```

knzo25 replied:
Addressed in [00d1a0c](https://github.com/tier4/aip_launcher/pull/348/commits/00d1a0c92eff6fd7c74cbf220c341a6eea8ad689)

mojomex commented on common_sensor_launch/launch/nebula_node_container.launch.py:
```diff
@@ -333,6 +406,9 @@ def add_launch_arg(name: str, default_value=None, description=None):
     add_launch_arg("use_multithread", "False", "use multithread")
     add_launch_arg("use_intra_process", "False", "use ROS 2 component container communication")
     add_launch_arg("lidar_container_name", "nebula_node_container")
+    add_launch_arg("pointcloud_container_name", "pointcloud_container")
+    add_launch_arg("use_pointcloud_container", "False")
+    add_launch_arg("use_cuda_preprocessor", "False")
     add_launch_arg("ptp_profile", "1588v2")
```
`lidar_container_name`, `pointcloud_container_name` and `use_pointcloud_container` could be confusing for users, how about for example `container_name` and `use_shared_container`, with the appropriate `description=`?

mojomex replied:
Also, are users always expected to use a shared container when using CUDA?

knzo25 replied:
Addressed in [411c39f](https://github.com/tier4/aip_launcher/pull/348/commits/411c39feaff2c906c06e60c194489a16a9adeeb2), 
315d2a4, and [f79d487](https://github.com/tier4/aip_launcher/pull/348/commits/f79d4875c315725de3562029a4904b548ac9b547).

I made it so that it was required (otherwise performance will not be good and I prefer it not working than working bad without the user realizing)

mojomex reviewed (commented):
Looking good! I left a few comments on readability/maintainability but that's all I think.
Is there a reason that x2(_gen2) is not covered?

knzo25 commented:
@mojomex 
Thanks for the review. X2 Gen2 was not covered on purpose since it is very difficult to maintain a compatible version of all the PRs (XX1 is better in this sense). As soon as the universe PRs get merged and I can get an easy way to port and test it on X2 Gen2 I will sent the proper PR here

knzo25 committed:
chore: refactored names

Signed-off-by: Kenzo Lobos-Tsunekawa <kenzo.lobos@tier4.jp>

knzo25 committed:
chore: refactored to simply branches

Signed-off-by: Kenzo Lobos-Tsunekawa <kenzo.lobos@tier4.jp>

knzo25 committed:
chore: added an assert to guarantee that the cuda preprocessor will be used with shared containers

Signed-off-by: Kenzo Lobos-Tsunekawa <kenzo.lobos@tier4.jp>

knzo25 committed:
chore: added documentation for the new arguments

Signed-off-by: Kenzo Lobos-Tsunekawa <kenzo.lobos@tier4.jp>

knzo25 committed:
chore: reimplementing the aip naming change instead of resolving conflicts

Signed-off-by: Kenzo Lobos-Tsunekawa <kenzo.lobos@tier4.jp>

knzo25 committed:
Merge branch 'tier4/universe' into feat/cuda_acceleration_and_transport_layer

Signed-off-by: Kenzo Lobos-Tsunekawa <kenzo.lobos@tier4.jp>

knzo25 committed:
chore: refactored node creation

Signed-off-by: Kenzo Lobos-Tsunekawa <kenzo.lobos@tier4.jp>

knzo25 commented:
@mojomex 
Addressed all comments. I can not ask for a review since the previous comments were not added in the context of a `require changes` it seems

mojomex commented on aip_common_sensor_launch/launch/nebula_node_container.launch.py:
```diff
@@ -336,7 +433,26 @@ def add_launch_arg(name: str, default_value=None, description=None):
     add_launch_arg("use_multithread", "False", "use multithread")
     add_launch_arg("use_intra_process", "False", "use ROS 2 component container communication")
-    add_launch_arg("lidar_container_name", "nebula_node_container")
+    add_launch_arg(
+        "lidar_container_name",
+        "nebula_node_container",
+        "Name of the new container to be created when use_shared_container is false",
+    )
+    add_launch_arg(
+        "shared_container_name",
+        "pointcloud_container",
+        "Name of the existing container to be used when use_shared_container is true",
+    )
```
These two should probably be combined for clarity. They can never validly be set at the same time and the shared container vs. new container aspect is already handled by the use_shared_container flag.

knzo25 replied:
Addressed in [fc2b7fd](https://github.com/tier4/aip_launcher/pull/348/commits/fc2b7fd2992092e5b43812fb42015bea96ec9492)
The consequence is that all containers will be called `pointcloud_container` which will need tooling that relied on this to be updated (not sure if the container name was ever used though)

mojomex reviewed (changes requested):
I'd like the two container_name params to be merged but otherwise LGTM

knzo25 committed:
feat: removed the shared_container_name. Now all containers will be called pointcloud_container. The difference being whether they are loaded in the global namespace or created in a new container within the lidar namespace

Signed-off-by: Kenzo Lobos-Tsunekawa <kenzo.lobos@tier4.jp>

knzo25 commented:
@mojomex 
Addressed the last comment.
Since we are close to merging this, I left the cuda acceleration and shared containers as disabled.
To enable them, you just need to change the following settings:

```xml
<arg name="use_shared_container" default="false"/>
<arg name="use_cuda_preprocessor" default="false"/>
```
to

```xml
<arg name="use_shared_container" default="true"/>
<arg name="use_cuda_preprocessor" default="true"/>
```

in `aip_xx1_launch/launch/lidar.launch.xml`

mojomex reviewed (approved):
[xx1-cuda.webm](https://github.com/user-attachments/assets/dd7bf091-4ab5-4ae4-a405-5733bd0cd454)

Works up until concat. I had to shoehorn the new autoware_universe into pilot-auto.xx1 so I could not test everything (at least autoware_tensorrt_plugins did not build successfully).

Since this is not this PR's fault, I'd say LGTM.

knzo25 committed:
Merge branch 'tier4/universe' into feat/cuda_acceleration_and_transport_layer
